### PR TITLE
Changed logging to be human readable.

### PR DESCRIPTION
**Pull request by spencerkimball**
Added color highlighting if the term supports it for warning=red,
error=bold&red.

We now build up a dictionary for output to a binary log (TODO).
I removed ClusterID from the log Field enum. It's completely superfluous.

Also changed the Infoc, Warningc, Errorc & Fatalc to take actual
format string and parameters. The other mechanism was just too clumsy
to be used easily. As part of the machine-readable dict, we output
the original format string and each of the parameters to it, along
with their types. I feel like this should allow pretty flexible
programmatic manipulation of logs.

Presumably the machine readable logs will be output as JSON and compressed.

### CODE REVIEW

File: util/log/clog.go [hunk 1]
```diff
@@ -705,6 +724,53 @@ func (l *loggingT) output(s severity, buf *buffer, file string, line int, alsoTo
...
+
+// checkForColorTerm attempts to verify that stderr is a character
+// device and if so, that the terminal supports color output.
+func (l *loggingT) checkForColorTerm() bool {
+	if l.hasColorTerm == nil {
+		var color bool
+		fi, _ := os.Stderr.Stat() // get the FileInfo struct describing the standard input.
+		if (fi.Mode() & os.ModeCharDevice) != 0 {
+			term := os.Getenv("TERM")
+			if strings.Contains(term, "xterm") && strings.Contains(term, "color") {
```
Comment (bdarnell) [inline util/log/clog.go]: I don't want to go too far down this rabbit hole, but TERM=ansi also seems to be somewhat standard and that's what my emacs shells use.

**Comment by bdarnell** [comment]
LGTM aside from the build failure.

**Comment by tschottdorf** [comment]
I'm critical about the strong preference for format+args over passing in key-value pairs. It means that whatever information you put in there is not available from the structured log output. I understand that especially now during development people need to actually read it, but I would prefer having a preference for fixed-string messages and contextualizing everything. Human-readable output could simply prefer the fields added on to the message (as opposed to spewing out the whole context) and the result should be what you intend without us having to enforce developer rigor for each and every log message in the future.

`log.Warningc(ctx, "async execution of failed: %s", err)`

Makes it really hard for me to learn anything about the error (this is from your change, btw).

`log.Warningc(ctx, "async execution failed", log.Error, err)`

is what I would like to enforce. Log output for humans could still be good with just a little bit of seasoning:

async execution failed: Error=...

So let's discuss whether we really want to go down this Printf-road. So far, I think it's a bad idea.

The logging code now also does more complicated stuff - reflection, creating a map for each request which it will have to sort before outputting - I suppose that could be ok if the performance works for us, but using a map for the key-value pairs means you're either sorting every time or you don't have order. Going back to me first point, it makes sense to have the separate inputs of key-value fields separated (from header, from context, from user kv-pairs). Maybe a slice is not ideal, but a map seems worse than something simple and slice-based. That's a minor point though.

And one more thing, if you're colorizing then you will also write colorized log files, which is likely not what you want. I had originally intended to use clog only for structured log and to simply print to Stderr from `log.go` instead for human logging, the idea being that you wouldn't ever want to log human output to files but maybe you would want human on stderr, structured in files.
We should figure out what we want and update the flags accordingly. I was going to do that, but you beat me to it.

File: util/log/clog.go (hunk 1 above)
Comment (spencerkimball) [inline util/log/clog.go]: How about a regexp?

`(ansi|xterm.*color)`

Comment (bdarnell) [inline util/log/clog.go]: Works for me.

**Comment by spencerkimball** [comment]
I'm going to rant for a second here: there are no machines parsing our
log files. We don't even have any plans on the near horizon to do
that. I can understand planning for the future, but I've been unable
to read the logs for more than a couple of weeks and that seems like a
really adverse outcome.

OK, to your points: the format + args strikes me as a suprisingly good
way to do things over having a single field for 'detail' and one for
'err', etc.

 - You may have more than a single detail or error. Both cases occur
   frequently. Are we going to add 'Detail2'? How about 'Detail3'?

 - I'm sure you know better than anyone how hard it is to write a
   log line where you need to assign things to fields. They just
   don't match up all that well. Or as in the previous point, you
   actually have two errors or two details. Which means you end up
   doing a fmt.Sprintf for the 'Msg' field. This is not just ugly
   and a pain, but you're immediately losing machine readability,
   which is the whole point!

 - The format string + machine readable args allows for comprehension
   not just when we're outputting the errors for humans. The detail
   and err being separate would make understanding them in the context
   of the disembodied message very hard (believe me, I really, really
   really tried--I just can't stress that enough).

 - The format string + machine readable args also allows for better
   comprehension for the machines! Well, not necessarily the machines
   themselves, but for the programmers who need to figure out just
   what in the heck a log line is supposed to mean. Especially when
   going back historically.

 - Format + args might not be as exacting as a strict schema for the
   different types of data which could be logged. I'm arguing this is
   a good thing because log messages don't / shouldn't follow a strict
   schema--the needs of a log line are pretty diverse and making each
   adhere to a schema feels to me like pushing a square peg into a
   round hole. I would also stress that we can write some simple
   utilities in our log processing tools to extract the schema you
   want pretty much automatically (that's why I included the types).

Yes, having the logs human readable is better during development.
That's why I'm going to insist we fix this now. But I would strongly
argue that they must always remain human readable. At least until
we're all replaced by artificial intelligences who can code circles
around us and can easily parse log lines which appear so repetitive
to my eyes, that they all look like exactly the same message.

> log.Warningc(ctx, "async execution of failed: %s", err)
>
> Makes it really hard for me to learn anything about the error (this is from your change, btw).

Wait, why is this hard to learn about the error? Are you referring
to reading the log file as it scrolls by, or trying to do something
programmatic with the log file? If programmatic, you can just use
the types to extract the errors from args. It's really not different
at all.

> async execution failed: Error=...
> 
> So let's discuss whether we really want to go down this Printf-road. So far, I think it's a bad idea.

One is easier to read, one not so much. Both are trivial to extract
the error for programmatic purposes.

Yes, this is going to potentially be inefficient. We can fix that.
The maps it's creating are very small, so will be trivial arrays in
Go. But that's immaterial: let's find the right logging solution
and then worry about performance.

> And one more thing, if you're colorizing then you will also write
> colorized log files, which is likely not what you want.

No, it only does the colorization on stderr.

We can probably write the json to the log file, but we'll need a
utility program to read the json-ified program and output proper
human readable log files. Not a problem with the way we're storing
all the info.

**Comment by tschottdorf** [comment]
> but I've been unable to read the logs for more than a couple of weeks and that seems like a
really adverse outcome.

I'm on your side with that one. Not suggesting that we should keep the way it is now.

> OK, to your points: the format + args strikes me as a suprisingly good way to do things over having a single field for 'detail' and one for 'err', etc.

That's not how it is, though. There can be arbitrary fields and specifying them is as easy as pie:

`log.Infof(ctx, "some message", "temperature", 90, "rant", true, "justified", "only somewhat")`

The fixed fields are only in the context. I agree that the Detail field there is nonsense. It's actually left from an earlier iteration (which might be the one that you're still thinking of in this rant).

I'm going to stick with my point that I think that the above is just as (or more legible as) the output of

`log.Infof(ctx, "some message with temperature %s and rant=%t and %s %s", temperature, 90, "rant", true, "justified", "only somewhat")`.

> "some message with temperature ......."

versus

> "some message: temperature=..., ....."

So, at least in my opinion, properly formatted for humans and colored, you have the same information nicely presented but you get for free that everything is compatible with structured logging.

I'm ignoring the other points because, as you say, they're less important now.

**Comment by spencerkimball** [comment]
They're really not as legible. I actually find bad english in the log lines incredibly difficult to read. We'd be institutionalizing broken English if we go down the path you're suggesting.

Perhaps a better way forward is to stick to a convention where we put the description of the argument before it, separated by `(\w*(:?|=?)\w*)` (sorry my regexp abilities are not great):

`log.Infoc(ctx, "some message with temperature %s and rant=%t and justified: %s", 90, true, "only somewhat")`

**Comment by tschottdorf** [comment]
I... think I don't really understand. Where is the broken english? I actually think inlining the information in the message so that you end up with something that passes as a sentence is annoying. It means that you might be printing the same information differently depending on your choice of %s, %v or the likes in that specified spot. I really just want a message telling me what goes on and then, next to it, the information that goes along with that message. Could you explain (sorry, one last time) where your aversion comes from? Maybe just pick a log message that annoys you in particular and we'll take a quick look at it.

**Comment by spencerkimball** [comment]
`log.Infof("splitting range %q-%q at keys %v", rng.Desc().StartKey, rng.Desc().EndKey, splitKeys)`
outputs: `I0521 01:18:34.091396   86388 split_queue.go:967] splitting range ""-"\xff\xff" at keys ["q"]`

`log.Infof("splitting range" "start", rng.Desc().StartKey, "end", rng.Desc().EndKey, "splits", splitKeys)`
outputs: `I0521 01:18:34.091396   86388 split_queue.go:967] splitting range: start="", end="\xff\xff", splits="q"`

We have to look at these things all day, in and out. There is a tax to the more structured output. I really think losing literate log statements is a step backwards.

**Comment by tschottdorf** [comment]
Honestly both options in that case are pretty much the same to me, and the second one has a lot of advantages. Sure, you're somewhat constrained in how you print the information locally, but you can change it at any given point uniformly across the codebase (aligned, multilined, different colors for the extra args, filterable, pretty-print logged structures, actually use the context for human output where you want it...). Whereas Printf for each statement is going to be a style circus with a lot of inconsistency (and it's never been great how we print for instance args and other structures in the middle of the sentence - our logging was pretty horrible prior to this change). Nor does it 
It's hard to believe that a hard-coded Sprintf beats every possible formatting of message and context. Don't you see a middle ground somewhere?

**Comment by spencerkimball** [comment]
You can do all of those things (not that I would advocate doing them, mind you) with either scheme. Let's be clear about the differences.

Printf-style
 - exact, utterly flexible human-readable format specifiable
 - relies on developer to put a helpful keyword next to format specifier if you want to reliably have an argument name

Structured-style
 - May structure as literate english, may not; you get a detail string and then a sequence of key/values, no matter which way you slice it
 - Always have name of argument

In both cases, we can output the things however we want. We have all of the same information, modulo not having a guaranteed argument name in one case, and not having a guaranteed literate / easy-to-grok english-language formulation on the other.

You say "hard coded sprintf" like it's a death sentence. As I'm pointing out, we get the exact same information, as in this case `log.Infoc(ctx, "some message with temperature %s and rant=%t and justified: %s", 90, true, "only somewhat")`.

My point is really that I want to read english, not line up all of the arguments or color code them. I actually think that would make them a lot harder to read, as was amply evidenced by the very regular output which the logs were exhibiting up to this change.

**Comment by mrtracy** [comment]
I guess the key decision here is whether or not *every* variable of every log message deserves to be part of the structured data, or if its okay for some variables to just be wrapped into the message.

If it's okay to structure only SOME variables, I would recommend adding them to the context before logging.

If we want *every* variable to be structured, I have a couple of alternatives:
**Printf style with names**
`log.Infoc("splitting range %q-%q at keys %v", "startKey", rng.Desc().StartKey, "endKey", rng.Desc().EndKey, "splitKeys", splitKeys)`

Where every variable has a title.  I mean it's a bit nasty but it gives you both a formatted message AND structured variables with keys.

**Alternate formatting structure**
`log.Infoc("splitting range {startKey}-{endKey} at keys {splitKeys}", "startKey", rng.Desc().StartKey, "endKey", rng.Desc().EndKey, "splitKeys", splitKeys)`

The log functions would add key/value pairs, and the message would reference them by key directly.  You could probably also reference keys from the context in the message.

**Comment by spencerkimball** [comment]
I did consider both of those options, but my issue is the same as you mention: it feels heavyweight. We could also do something like:

`log.Infoc("splitting range {start}%q-{end}%q at keys %v", rng.Desc().StartKey, rng.Desc().EndKey, splitKeys)`

Notice that we only specify `{start}` and `{end}` to fill in the missing argument names, but don't have to do anything else differently.

One thing to keep in mind here is whether it's a good idea to do something different with our logging than the other 99% of projects (printf style is pretty universal). I'd say there would have to be a good reason to do that and I'm failing to see enough of a justification.

**Comment by spencerkimball** [comment]
Your point about structuring only some of the vars is a good one, but I think augmenting the context is going to be ugly in situ.

**Comment by tschottdorf** [comment]
Yeah, it shouldn't feel heavyweight. I'm just really happy with something like this (logrus, by the way popular and structured).



I don't need my logging in prose. I want the information in a concise way. The argument that the current state of the logging proves that the other option is better is pretty invalid: The current logging is our old logging, only with more trash - we don't currently have real structured logging, especially not optimized for human consumption.

I'm not sure what else I can say. It seems like an obvious choice for me, but I suppose it's the same for you. I'd still roll with structured-only, focus on clean presentation and just use Sprintf manually in the .1% of cases where nothing else works.

**Comment by spencerkimball** [comment]
Feel free to play with structured. If you can make real log output look as clear as the logrus example, I'd be all for it. But I think logrus isn't going to look like that with more reasonable production logs.

I don't agree that our logging is terrible because it isn't "structured". It is actually quite structured--we know the types, values, and can surmise nearly all of the argument names. There are other reasons it's not so great.

I'm going to check this change in though, as I think we need to restore visibility to our logs.

**Comment by tschottdorf** [comment]
ok, LMTM (looks mergable to me :) )

**Pull request merged**